lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -1150,6 +1191,180 @@ void SESERegionBuilder::ensureSingleExitFromLoops() {
+  // preheader.
+  SILBasicBlock *newHeader = loop->getHeader();
+  auto preheaderTermInst = dyn_cast<BranchInst>(preheader->getTerminator());
+  assert(preheaderTermInst && "Preheader of a loop has a non-branch terminator");
+  for (unsigned argIndex = 0; argIndex < oldHeaderNumArgs; ++argIndex) {
+    auto preheaderArg = preheaderTermInst->getArg(argIndex);
+    auto newHeaderArg = newHeader->getArgument(argIndex);
+    cloner.updateValueMap(newHeaderArg, preheaderArg);
+  }
+  // Clone everything except the new header. We should traverse the
+  // blocks in depth first order to ensure values are cloned before they are used.
+  SmallPtrSet<SILBasicBlock *, 32> worklist;
+  SmallVector<SILBasicBlock *, 32> initializedBlocks;
+  worklist.insert(header);
+  while (!worklist.empty()) {
+    SILBasicBlock *current = *worklist.begin();
+    worklist.erase(current);
+    cloner.initBlock(current);
+    initializedBlocks.push_back(current);
+    for (SILBasicBlock *succ : current->getSuccessorBlocks()) {
+      // Skip if succ is not a part of the loop, is already cloned, or
+      // is the new preheader.
+      if (!loop->contains(succ) || cloner.remapBasicBlock(succ) != succ ||
+          succ == newHeader) {
+        continue;
+      }
+      worklist.insert(succ);
+    }
+  }
+
+  SILLoop *parentLoop = loop->getParentLoop();
+  for (SILBasicBlock *bb : initializedBlocks) {
+    cloner.cloneBlock(bb);
+    SILBasicBlock *clonedBlock = cloner.cloneBlock(bb);
+    if (parentLoop) {
+      parentLoop->addBasicBlockToLoop(clonedBlock, LI->getBase());
+    }
+  }
+
+  // Get the clone for the original and new header.
+  SILBasicBlock *clonedHeader = cloner.remapBasicBlock(header);
+  replaceBranchTarget(preheader->getTerminator(), newHeader, clonedHeader,
+                      /*preserveArgs*/ false);
+
+  // (2) Patch values that still refer to the arguments of `newHeader`.
+  //
+  SILBasicBlock *newLatch = loop->getLoopLatch();
+  SILBasicBlock *clonedNewLatch = cloner.remapBasicBlock(newLatch);
+  for (SILBasicBlock *pred : newLatch->getPredecessorBlocks()) {
+    auto predTermInst = dyn_cast<BranchInst>(pred->getTerminator());
+    assert(predTermInst && "Preheader of a loop has a non-branch terminator");
+    for (unsigned argIndex = 0; argIndex < predTermInst->getNumArgs(); ++argIndex) {
+      auto arg = predTermInst->getArg(argIndex);
+      // Skip if this is not an argument of the `newHeader`.
+      if (!isa<SILArgument>(arg) ||
+          cast<SILArgument>(arg)->getParent() != newHeader) {
+        continue;
+      }
+      // Iterate over the incoming values of the corresponding argument in the
+      // latch block and pick one that is suitable to be used here.
+      auto destBBArg = newLatch->getArgument(argIndex);
+      SmallVector<SILValue, 8> incomingValues;
+      destBBArg->getIncomingPhiValues(incomingValues);
+      for (auto value : incomingValues) {
+        if (value != arg && DI->properlyDominates(value, predTermInst)) {
```

Comment:
what if we cannot find such a value? should we assert this won't ever happen?

Reply:
It is guaranteed to find a value. SIL verification will fail otherwise. No need to add another check that will essentially replicate SIL verification.

Reply:
Debugging SIL verifier failure usually takes more work because the context is non-local. It'd usually be preferrable to have a local check so that we can fail fast. Also, the check serves as a documentation for this important invariant.

Reply:
Good point. Added an assert to check that we patch such arguments.